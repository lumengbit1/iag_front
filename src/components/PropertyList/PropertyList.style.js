import styled, { css } from 'styled-components';

const StyledButton = css`
  position: absolute;
  left: 50%;
  bottom: 1%;
  transform: translate(-50%, -50%);
  width: 180px;
  height: 20px;
  border-radius: 10px;
  display: none;

  &:hover {
    cursor: pointer;
  }
`;

export const Button = styled.button`
  ${StyledButton};

  background-color: #f3c1c2;
  border: 1px solid #cc0007;
  color: #a21519;

  ${(props) => props.isResult && css`
    background-color: #e3f1df;
    border: 1px solid #4d8437;
    color: #4b8333;
  `}
`;

export const PropertyContainer = styled.div`
  display: flex;
  flex-direction: column;
  position: relative;
  width: 200px;
  height: 180px;
  border: 1px solid #000;
  border-radius: 10px;
  margin: 10px;

  &:hover {
    button {
      display: block;
    }
  }
`;
