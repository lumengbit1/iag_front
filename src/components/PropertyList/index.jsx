import React from 'react';
import PropTypes from 'prop-types';
import ImmutablePropTypes from 'react-immutable-proptypes';
import withComponentLoading from './util';
import Property from '../Property';
import {
  Button,
  PropertyContainer,
} from './PropertyList.style';

const PropertyList = (props) => {
  const { records, btnText, isResult, clickAction } = props;

  return (
    <>
      {records.map((item) => (
        <PropertyContainer key={item.get('id')} data-testid="property">
          <Property
            id="property"
            price={item.get('price')}
            color={item.getIn(['agency', 'brandingColors', 'primary'])}
            logo={item.getIn(['agency', 'logo'])}
            mainImage={item.get('mainImage')}
          />
          <Button
            data-testid="test"
            isResult={isResult}
            onClick={() => clickAction(item.get('id'))}
          >
            {btnText}
          </Button>
        </PropertyContainer>
      ))}
    </>
  );
};

PropertyList.propTypes = {
  records: ImmutablePropTypes.list.isRequired,
  btnText: PropTypes.string.isRequired,
  isResult: PropTypes.bool.isRequired,
  clickAction: PropTypes.func.isRequired,
};

export default withComponentLoading(PropertyList);
