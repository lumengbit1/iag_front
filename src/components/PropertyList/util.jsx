import React from 'react';
import PropTypes from 'prop-types';
import { useDispatch, useSelector } from 'react-redux';
import { getResultsAction, getSavedAction, add_property, remove_property } from '../../reducers/actions';

function componentType(type) {
  if (type === 'results') {
    return {
      clickAction: add_property,
      actionFunction: getResultsAction,
      btnText: 'Add Property',
      isResult: true,
    };
  }

  return {
    clickAction: remove_property,
    actionFunction: getSavedAction,
    btnText: 'Remove Property',
    isResult: false,
  };
}

const withComponentLoading = (WrappedComponent) => {
  function WithWrap(props) {
    const { type } = props;
    const dispatch = useDispatch();

    const value = useSelector((state) => state.getIn(['value', type]));
    const loading = useSelector((state) => state.getIn(['value', 'loading', type]));

    React.useEffect(() => {
      dispatch(componentType(type).actionFunction());
    }, []);

    if (loading) return <div data-testid="loading">Loading...</div>;

    if (!value || value.isEmpty()) return null;

    return (
      <WrappedComponent
        records={value}
        isResult={componentType(type).isResult}
        clickAction={(params) => dispatch(componentType(type).clickAction(params))}
        btnText={componentType(type).btnText}
      />
    );
  }

  WithWrap.propTypes = {
    type: PropTypes.string.isRequired,
  };

  return WithWrap;
};

export default withComponentLoading;
