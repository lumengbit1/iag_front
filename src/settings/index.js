export default {
  RESULTS_BASE_API_DOMAIN: 'http://localhost:3000/results',
  SAVED_BASE_API_DOMAIN: 'http://localhost:3000/saved',
};
