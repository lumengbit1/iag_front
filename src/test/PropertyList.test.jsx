import React from 'react';
import 'jest-styled-components';
import axios from 'axios';
import * as redux from 'react-redux';

import { render, cleanup, waitFor, fireEvent } from '@testing-library/react';
import thunk from 'redux-thunk';
import MockAdapter from 'axios-mock-adapter';
import { createStore, applyMiddleware } from 'redux';
import { combineReducers } from 'redux-immutable';
import PropertyList from '../components/PropertyList';
import eventList from '../reducers/reducer';
import settings from '../settings';

const store = createStore(combineReducers({ value: eventList }), applyMiddleware(thunk));

afterEach(cleanup);

const mockResultsData = [
  {
    price: '$726,500',
    agency: {
      brandingColors: {
        primary: '#ffe512',
      },
      logo: 'https://i1.au.reastatic.net/170x32/d9e65c666e427e655fb63dcfe73f50d4ac7ff9a58c173db9474bd92e75b01070/main.gif',
    },
    id: '1',
    mainImage: 'https://i2.au.reastatic.net/640x480/20bfc8668a30e8cabf045a1cd54814a9042fc715a8be683ba196898333d68cec/main.jpg',
  },

];

const mockSavedData = [
  {
    price: '$526,500',
    agency: {
      brandingColors: {
        primary: '#000000',
      },
      logo: 'http://i2.au.reastatic.net/agencylogo/WVYSSK/2/20140701084436.gif',
    },
    id: '4',
    mainImage:
      'http://i2.au.reastatic.net/640x480/5e84d96722dda3ea2a084d6935677f64872d1d760562d530c3cabfcb7bcda9c2/main.jpg',
  },
];

describe('Render Test', () => {
  it('1: expect rendering correct when type=results', () => {
    const { container } = render(
      <redux.Provider store={store}>
        <PropertyList type="results" />
      </redux.Provider>,
    );
    expect(container).toMatchSnapshot();
  });

  it('2: expect rendering correct when type=saved', () => {
    const { container } = render(
      <redux.Provider store={store}>
        <PropertyList type="saved" />
      </redux.Provider>,
    );
    expect(container).toMatchSnapshot();
  });
});

describe('Function Test', () => {
  it('1: expect getResultsAction becalled', async () => {
    const mock = new MockAdapter(axios);
    mock.onGet(`${settings.RESULTS_BASE_API_DOMAIN}`).reply(200, mockResultsData);

    const { getByTestId } = render(
      <redux.Provider store={store}>
        <PropertyList type="results" />
      </redux.Provider>,
    );

    await waitFor(() => getByTestId('property'));

    expect(getByTestId('property')).toBeInTheDocument();
  });

  it('2: expect getSavedAction becalled', async () => {
    const mock = new MockAdapter(axios);
    mock.onGet(`${settings.SAVED_BASE_API_DOMAIN}`).reply(200, mockSavedData);

    const { getByTestId } = render(
      <redux.Provider store={store}>
        <PropertyList type="saved" />
      </redux.Provider>,
    );

    await waitFor(() => getByTestId('property'));

    expect(getByTestId('property')).toBeInTheDocument();
  });

  it('3: expect results button onclick event', async () => {
    const mock = new MockAdapter(axios);
    mock.onGet(`${settings.RESULTS_BASE_API_DOMAIN}`).reply(200, mockResultsData);
    const useDispatchSpy = jest.spyOn(redux, 'useDispatch');
    const mockDispatchFn = jest.fn();
    useDispatchSpy.mockReturnValue(mockDispatchFn);
    const { getByTestId } = render(
      <redux.Provider store={store}>
        <PropertyList type="results" />
      </redux.Provider>,
    );
    await waitFor(() => getByTestId('test'));
    fireEvent.click(getByTestId('test'));
    expect(mockDispatchFn).toBeCalled();
  });
});
