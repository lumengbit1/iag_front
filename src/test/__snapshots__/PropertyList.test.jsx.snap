// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Test 1: expect rendering correct when type=results 1`] = `
<div>
  <div
    data-testid="loading"
  >
    Loading...
  </div>
</div>
`;

exports[`Render Test 2: expect rendering correct when type=saved 1`] = `
<div>
  <div
    data-testid="loading"
  >
    Loading...
  </div>
</div>
`;
